File: algorithms/java/Missing Number.java
public class Solution {
    private void swap(int[] nums, int i, int j) {
        int tmp = nums[i];
        nums[i] = nums[j];
        nums[j] = tmp;
    }
    
    public int missingNumber(int[] nums) {
        int n = nums.length;
        int i = 0;
        while(i < nums.length) {
            if(nums[i] >= n) {
                i++;
            } else {
                if(nums[i] != i) {
                    swap(nums, i, nums[i]);
                } else{
                    i++;
                }
            }
        }
        
        for(int j = 0;j < n;++j) {
            if(j != nums[j]) {
                return j;
            }
        }
        
        return n;
    }
}

File: algorithms/cpp/Missing Number.cpp
class Solution {
public:
    int missingNumber(vector<int>& nums) {
        nums.push_back(-1);
        for(int i = 0;i < nums.size();++i) {
            while(nums[i] != i && nums[i] != -1) {
                swap(nums[i], nums[nums[i]]);
            }
        }
        
        for(int i = 0;i < nums.size();++i) {
            if(nums[i] != i) return i;
        }
    }
};